### PR TITLE
[SPARK-14571][ML]Log instrumentation in ALS

### PR DESCRIPTION
**Pull request by wangmiao1981**
## What changes were proposed in this pull request?

Add log instrumentation for parameters:
rank, numUserBlocks, numItemBlocks, implicitPrefs, alpha,
userCol, itemCol, ratingCol, predictionCol, maxIter,
regParam, nonnegative, checkpointInterval, seed

Add log instrumentation for numUserFeatures and numItemFeatures
## How was this patch tested?

Manual test: Set breakpoint in intellij and run def testALS(). Single step debugging and check the log method is called.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 1]
```diff
@@ -395,13 +395,21 @@ class ALS(@Since("1.4.0") override val uid: String) extends Estimator[ALSModel]
...
+                    userCol, itemCol, ratingCol, predictionCol, maxIter,
+                    regParam, nonnegative, checkpointInterval, seed)
     val (userFactors, itemFactors) = ALS.train(ratings, rank = $(rank),
       numUserBlocks = $(numUserBlocks), numItemBlocks = $(numItemBlocks),
       maxIter = $(maxIter), regParam = $(regParam), implicitPrefs = $(implicitPrefs),
       alpha = $(alpha), nonnegative = $(nonnegative),
       checkpointInterval = $(checkpointInterval), seed = $(seed))
     val userDF = userFactors.toDF("id", "features")
     val itemDF = itemFactors.toDF("id", "features")
+    val numUserFeatures = userDF.collect.size
```
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: I don't think we need to collect the dataframes locally here to get the counts (maybe just use `count`) or maybe we could put this logging somewhere that already have the counts?

Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: We definitely don't want to collect.

I don't see that we can put the logging in say `ALS.train` without changing the signature to take the `instr` instance. I don't think we want to do this.

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 2]
```diff
@@ -395,13 +395,21 @@ class ALS(@Since("1.4.0") override val uid: String) extends Estimator[ALSModel]
...
+    val numItemFeatures = itemDF.collect.size
+    instr.logNumFeatures(numUserFeatures)
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: In the instrumenter, this does `log(compact(render("numFeatures" -> num)))` so the message will be the same with different numbers, which is confusing as you then don't know which is user and item. I think we need a custom message here.

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala (hunk 1 above)
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: +1. The instrumentation should not make the current algorithm slower:
 - it should not require evaluating more RDDs / DFs
 - it should not require a collect() call

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala (hunk 2 above)
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: +1. It seems worth to add a method in the instrumentation that does something like that:

```scala
  def logValue(name: String, num: Long): Unit = {
    log(compact(render(name -> num)))
  }
```

**Comment by thunterdb** [comment]
@wangmiao1981 thanks for the PR. Can you see if you could make it work without having to evaluate the input data?

**Comment by wangmiao1981** [comment]
Thanks all for your comments! Let me figure out how to collect the information without slowing the algorithm. @MLnick The names are passed to the log. For example, numUserItems will be logged as "numUserItems 20". The core issue here is not collecting the data.

**Comment by wangmiao1981** [comment]
@thunterdb train method has count information, but it will change the signature of the train method. I am learning how to avoid collect and changing signature. Thanks all for review.

**Comment by wangmiao1981** [comment]
retest this please

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 3]
```diff
@@ -607,7 +611,8 @@ object ALS extends DefaultParamsReadable[ALS] with Logging {
       intermediateRDDStorageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK,
       finalRDDStorageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK,
       checkpointInterval: Int = 10,
-      seed: Long = 0L)(
+      seed: Long = 0L,
+                           instr: Option[Instrumentation[ALS]] = None)(
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: We should not change the ```train``` API, it will break binary compatibility.

Comment (wangmiao1981) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: Modifying training is to get the number of features without collect. I agree with you that we should not change train. But I can't get number of features without changing it.

**Comment by wangmiao1981** [comment]
@MLnick Yanbo does not like the change of train() API. The new parameter is optional, so the user of train should not be aware of this change. In addition, I think ML user should use fit instead of using train directly. I am not sure how the Developer API is used by developers. Any comments/suggestions?

**Comment by MLnick** [comment]
@wangmiao1981 I tend to agree with @yanboliang. I don't think adding instrumentation is critical enough to break the `train` method signature, even if it is `DeveloperApi`.

I think we should just log what we can unobtrusively, and perhaps at a later stage further instrumentation can be added in a performant way.

**Comment by wangmiao1981** [comment]
@MLnick I agree. I will remove the feature log now and only log parameters. I will keep the named feature method.

**Comment by thunterdb** [comment]
@wangmiao1981 sorry for the delay. I agree with @MLnick that the developer API of ALS is already brittle enough as it stands. One way to still do it is to add one layer of indirection:

```scala
  private def train0[ID: ClassTag](
      ratings: RDD[Rating[ID]],
      rank: Int,
      numUserBlocks: Int,
      numItemBlocks: Int,
      maxIter: Int,
      regParam: Double,
      implicitPrefs: Boolean,
      alpha: Double,
      nonnegative: Boolean,
      checkpointInterval: Int,
      seed: Long,
      instrumentation: Instrumentation[ALS])(
      implicit ord: Ordering[ID]) = (RDD[(ID, Array[Float])], RDD[(ID, Array[Float])]) = { // all the code here}

  @DeveloperApi
  def train[ID: ClassTag]( // scalastyle:ignore
      ratings: RDD[Rating[ID]],
      rank: Int = 10,
      numUserBlocks: Int = 10,
      numItemBlocks: Int = 10,
      maxIter: Int = 10,
      regParam: Double = 1.0,
      implicitPrefs: Boolean = false,
      alpha: Double = 1.0,
      nonnegative: Boolean = false,
      intermediateRDDStorageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK,
      finalRDDStorageLevel: StorageLevel = StorageLevel.MEMORY_AND_DISK,
      checkpointInterval: Int = 10,
      seed: Long = 0L)(
      implicit ord: Ordering[ID]): (RDD[(ID, Array[Float])], RDD[(ID, Array[Float])]) = train0(
        ratings = ratings, 
        ...)
```

This minimizes the changes and keeps the API binary compatible, however it is not pretty. @MLnick @yanboliang  what do you think? I do not have a strong opinion on this one.

**Comment by thunterdb** [comment]
In any case, we should also log the success with `instrLog.logSuccess(model)`

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 4]
```diff
@@ -395,6 +395,10 @@ class ALS(@Since("1.4.0") override val uid: String) extends Estimator[ALSModel]
       .map { row =>
         Rating(row.getInt(0), row.getInt(1), row.getFloat(2))
       }
+    val instrLog = Instrumentation.create(this, ratings)
+    instrLog.logParams(rank, numUserBlocks, numItemBlocks, implicitPrefs, alpha,
+                    userCol, itemCol, ratingCol, predictionCol, maxIter,
```
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: indentation

**Comment by wangmiao1981** [comment]
@MLnick @yanboliang  Any further comments? Thanks!

**Comment by MLnick** [comment]
I'd go with what we have here for now. We could revisit later if necessary. If a user really wants to check the number of user/item factors they can do a count on the DFs themselves

**Comment by wangmiao1981** [comment]
@thunterdb What do you think about our discussions? Thanks!

Miao

**Comment by yanboliang** [comment]
This looks good to me, thanks!

File: mllib/src/main/scala/org/apache/spark/ml/util/Instrumentation.scala [hunk 5]
```diff
@@ -85,6 +85,13 @@ private[ml] class Instrumentation[E <: Estimator[_]] private (
   }
 
   /**
+   * Logs the value with customized name field.
+   */
+  def logNamedValue(name: String, num: Long): Unit = {
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/util/Instrumentation.scala]: Just a note for posterity, I thought about whether this should be a Double to support arbitrary values (maybe in future we want to log some training metrics such as log loss or other error, for example), but then the formatting would be off for integers. So if we need that in future we can just create an overloaded method.

**Comment by MLnick** [comment]
LGTM. Merged to master. Thanks!

**Pull request closed without merging**